README.md
[中文文档](https://github.com/booyangcc/excelstructure/blob/main/README_zh.md)

## Introduction

excelstructure is a tool for converting between Excel and Go struct. It can convert a struct to Excel and vice versa. It also supports getting the value of a single field from a row.

## Basic Usage

### Installation

`go get github.com/booyangcc/excelstructure`

### Example

```golang
package main

import (
	"fmt"
	"github.com/booyangcc/utils/convutil"
    "github.com/booyangcc/excelstructure"
)

type Info struct {
	Name    string  `excel:"column:user_name;comment:person name"`
	Phone   *string `excel:"column:phone;comment:phone number"`
	Age     string  `excel:"column:age;"`
	Man     bool    `excel:"column:man;default:true"`
	Address string  `excel:"column:address;skip"` // skip this field
}

var (
	infos = []*Info{
		{
			Name:    "booyang",
			Phone:   convutil.String("123456789"),
			Age:     "18",
			Man:     true,
			Address: "beijing",
		},
		{
			Name:    "booyang1",
			Phone:   convutil.String("123456789"),
			Age:     "14",
			Man:     false,
			Address: "shanghai",
		},
	}
)

func TestWriteRead() {
	p := excelstructure.NewParser("./test_write.xlsx")
	err := p.Write(infos)
	if err != nil {
		fmt.Println(err)
	}

	// Since the struct being written has a comment field, the second row in the Excel file will be a comment, so the data index offset is 2
	p.DataIndexOffset = 2
	var newInfo []*Info
	err = p.Read(&newInfo)
	if err != nil {
		fmt.Println(err)
	}
	fmt.Printf("oldInfo: %+v, newInfo: %+v", infos, newInfo)
}

func TestParse() {
	r := excelstructure.NewParser("./test_write.xlsx")
	// Since the struct being written has a comment field, the second row in the Excel file will be a comment, so the data index offset is 2
	r.DataIndexOffset = 2
	excelData, err := r.Parse()
	if err != nil {
		fmt.Println(err)
	}

	s := excelData.SheetNameData["Infos"]
	row3UserName, err := s.GetStringValue(3, "user_name")
	if err != nil {
		fmt.Println(err)
	}

	row3age, err := s.GetIntValue(3, "age")
	if err != nil {
		fmt.Println(err)
	}
	fmt.Println(row3UserName, row3age)

	row4UserName, err := s.GetStringValue(4, "user_name")
	if err != nil {
		fmt.Println(err)
	}

	row4age, err := s.GetIntValue(4, "age")
	if err != nil {
		fmt.Println(err)
	}
	fmt.Println(row4UserName, row4age)
}

```
Usage
### Tag Configuration
The tag name is `excel`. The tag field configuration is as follows:
> Name string `excel:"column:user_name;comment:person name;skip;default:boo"`
- column: the header to parse or write to Excel
- comment: if any field in the struct contains this configuration in the excel tag, the second row of the output Excel file will be a comment
- skip: indicates that the current field is skipped and not parsed or written to Excel
- default: if the field is zero-value, use the default value instead

### Parser Usage
Parser parameters:
- FileName: the file to read
- DataIndexOffset: the data index offset. If the first row is a header, the offset is 1. If there is a comment occupying a row, the offset is 2. The default value is 1.
- BoolTrueValues: the optional values for true boolean values. The default values are [true,True,TRUE,1,是,yes,Yes,YES,y,Y]. You can specify them manually.
- IsCheckEmpty: whether to check for empty values when serializing to a struct. If a value is empty, an error is thrown.
- IsEmptyFunc: the callback function to check for empty values. The default function is func(v string) bool {return v==""}
- IsCoordinatesABS: the type of cell coordinate value. If true, the coordinate is A1. If false, the coordinate is 1.
- ExcelData: the parsed data values
- AllowFieldRepeat: whether to allow duplicate fields. If true, the fields will be overwritten.
### Parse Struct to Excel
```golang
p := excelstructure.NewParser("./test_write.xlsx")
err := p.Write(infos)  
if err != nil {  
    fmt.Println(err)  
}  
```
### Parsing Excel to Struct
#### Method 1: Parse Directly to Struct
```golang
p := NewParser("./test_write.xlsx")  
// because the struct field has comment tag, so the comment has been written to row 2, so when read, data offset is 2
p.DataIndexOffset = 2
var newInfo []*Info
err = p.Read(&newInfo)
if err != nil {
        fmt.Println(err)
}
```
#### Method 2: Get the Value of a Single Field in a Row
```golang
p := NewParser("./test_write.xlsx")
// Since the struct being written has a comment field, the second row in the Excel file will be a comment, so the data index offset is 2
p.DataIndexOffset = 2  
excelData, err := p.Parse()  
if err != nil {  
fmt.Println(err)  
}  
  
s := excelData.SheetNameData["Infos"]  
row3UserName, err := s.GetStringValue(3, "user_name")  
if err != nil {  
fmt.Println(err)  
}  
  
row3age, err := s.GetIntValue(3, "age")  
if err != nil {  
fmt.Println(err)  
}  
fmt.Println(row3UserName, row3age)
```







CHANGELOG.md
## [v0.0.1] - 2023-08-18
- added convert struct to excel
- added convert excel to struct
- refactor code


parse_test.go
package excel

import (
	"fmt"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

type Info struct {
	Name    string  `excel:"column:user_name;"`
	Phone   *string `excel:"column:phone;"`
	Age     string  `excel:"column:age;"`
	Man     bool    `excel:"column:man;default:true"`
	Address string  `excel:"column:address;skip"` // skip this field
}

func TestParser_UnmarshalWithSheetIndex(t *testing.T) {
	var info []*Info
	p := NewParser("./test_excel_file/test.xlsx")
	err := p.UnmarshalWithSheetIndex(2, &info)
	if err != nil {
		assert.Error(t, err)
	}
	require.Equal(t, 4, len(info))
	require.Equal(t, "booyang1", info[0].Name)
}

func TestParser_Unmarshal(t *testing.T) {
	var info []*Info
	p := NewParser("./test_excel_file/test.xlsx")
	err := p.Unmarshal(&info)
	if err != nil {
		assert.Error(t, err)
	}
	require.Equal(t, 4, len(info))
	require.Equal(t, "booyang", info[0].Name)
}

// excel data index offset is 2, the first two rows are not data,
// first row is title,second row is comment
func TestParser_UnmarshalWithComment(t *testing.T) {
	var info []*Info
	p := NewParser("./test_excel_file/test_with_comment.xlsx")
	p.DataIndexOffset = 2
	err := p.Unmarshal(&info)
	if err != nil {
		assert.Error(t, err)
	}
	require.Equal(t, 4, len(info))
	require.Equal(t, "booyang", info[0].Name)
}

func TestParser_UnmarshalWithCheckEmpty(t *testing.T) {
	var info []*Info
	p := NewParser("./test_excel_file/test_check_empty.xlsx")
	p.IsCheckEmpty = true
	err := p.Unmarshal(&info)
	// one row is empty, so the error is not nil, "C3 field age value is empty"
	assert.Error(t, err)
	// length is 4, but one row is empty, so the length is 3
	require.Equal(t, 3, len(info))
	require.Equal(t, "booyang", info[0].Name)
}

func TestSheetData_GetIntValue1(t *testing.T) {
	p := NewParser("./test_excel_file/test.xlsx")
	data, err := p.Parse()
	if err != nil {
		fmt.Println(err)
	}
	s := data.SheetIndexData[1]
	row2, err := s.GetIntValue(2, "age")
	assert.NoError(t, err)
	require.Equal(t, 20, row2)

	row3, err := s.GetIntValueWithMultiError(3, "age", err)
	assert.NoError(t, err)
	require.Equal(t, 3, row3)
}


reader_test.go
package excelstructure

import (
	"fmt"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestParser_ReadWithSheetIndex(t *testing.T) {
	var info []*Info
	p := NewParser("./test_excel_file/test.xlsx")
	err := p.ReadWithSheetName("Sheet1", &info)
	if err != nil {
		assert.Error(t, err)
	}
	require.Equal(t, 4, len(info))
	require.Equal(t, "booyang1", info[0].Name)
}

func TestParser_Read(t *testing.T) {
	var info []*Info
	p := NewParser("./test_excel_file/test.xlsx")
	err := p.Read(&info)
	if err != nil {
		assert.Error(t, err)
	}
	require.Equal(t, 4, len(info))
	require.Equal(t, "booyang", info[0].Name)
}

// excel data index offset is 2, the first two rows are not data,
// first row is title,second row is comment
func TestParser_ReadWithComment(t *testing.T) {
	var info []*Info
	p := NewParser("./test_excel_file/test_with_comment.xlsx")
	p.DataIndexOffset = 2
	err := p.Read(&info)
	if err != nil {
		assert.Error(t, err)
	}
	require.Equal(t, 4, len(info))
	require.Equal(t, "booyang", info[0].Name)
}

func TestParser_ReadWithCheckEmpty(t *testing.T) {
	var info []*Info
	p := NewParser("./test_excel_file/test_check_empty.xlsx")
	p.IsCheckEmpty = true
	err := p.Read(&info)
	// one row is empty, so the error is not nil, "C3 field age value is empty"
	assert.Error(t, err)
	// length is 4, but one row is empty, so the length is 3
	require.Equal(t, 3, len(info))
	require.Equal(t, "booyang", info[0].Name)
}

func TestSheetData_GetIntValue1(t *testing.T) {
	p := NewParser("./test_excel_file/test.xlsx")
	data, err := p.Parse()
	if err != nil {
		fmt.Println(err)
	}
	s := data.SheetIndexData[1]
	row2, err := s.GetIntValue(2, "age")
	assert.NoError(t, err)
	require.Equal(t, 20, row2)

	row3, err := s.GetIntValueWithMultiError(3, "age", err)
	assert.NoError(t, err)
	require.Equal(t, 3, row3)
}


writer_test.go
package excelstructure

import (
	"github.com/booyangcc/utils/convutil"
	"github.com/stretchr/testify/assert"
	"testing"
)

func TestWriter_Marshal(t *testing.T) {
	w := NewParser("./test_excel_file/test_write.xlsx")
	infos := []*Info{
		{
			Name:    "booyang",
			Phone:   convutil.String("123456789"),
			Age:     "18",
			Man:     true,
			Address: "beijing",
		},
		{
			Name:    "booyang1",
			Phone:   convutil.String("123456789"),
			Age:     "14",
			Man:     false,
			Address: "shanghai",
		},
	}
	err := w.Write(infos)
	assert.NoError(t, err)
}


err.go
package excelstructure

import (
	"errors"
	"fmt"
)

var (
	// ErrorDataRowOutOfRange data row out of range
	ErrorDataRowOutOfRange = errors.New("out of range")
	// ErrorRowIndexIsHeader row index is header
	ErrorRowIndexIsHeader = errors.New("row index must > 1, because row 1 is header")
	// ErrorSheetName sheet name invalid
	ErrorSheetName = errors.New("sheet name not exist")
	// ErrorInOutputType output type invalid
	ErrorInOutputType = errors.New("output type must be slice or struct pointer")
	// ErrorSliceElemType elem type must struct or struct pointer
	ErrorSliceElemType = errors.New("slice elem type must struct or struct pointer")
	// ErrorTypePointer output not pointer
	ErrorTypePointer = errors.New("output type must be pointer")
	// ErrorNoSheet no sheet
	ErrorNoSheet = errors.New("no sheet")

	// ErrorFieldNotExist Field invalid
	ErrorFieldNotExist = errors.New("field not exist in header")
	// ErrorFieldRepeat repeat field
	ErrorFieldRepeat = errors.New("field repeat")
	// ErrorFieldTypeNotSupport field type not support
	ErrorFieldTypeNotSupport = errors.New("field type not support")
	// ErrorFieldNotMatch Field type not match
	ErrorFieldNotMatch = errors.New("field tag default value or excel value not match struct field type")
	// ErrorFieldNotSet pointer field can not set
	ErrorFieldNotSet = errors.New("pointer field can not set")
	// ErrorFieldInvalid pointer field invalid
	ErrorFieldInvalid = errors.New("pointer or field invalid")
	// ErrorFieldValueEmpty  field empty
	ErrorFieldValueEmpty = errors.New("value is empty")
	// ErrorNoData no data
	ErrorNoData = errors.New("no data")
)

// Error excel structure error
type Error struct {
	FileName    string
	SheetName   string
	Coordinates string
	Err         error
}

// NewError new error
func NewError(filName string, sheetName string, coordinates string, err error) error {
	return &Error{
		FileName:    filName,
		SheetName:   sheetName,
		Err:         err,
		Coordinates: coordinates,
	}
}

// Error error
func (e *Error) Error() string {
	return fmt.Sprintf("FileName: %s, SheetName: %s, Coordinates: %s , ErrMsg: %s",
		e.FileName, e.SheetName, e.Coordinates, e.Err.Error())
}


parse.go
package excelstructure

import (
	"github.com/xuri/excelize/v2"
)

var boolTrueValue = []string{
	"true",
	"True",
	"TRUE",
	"1",
	"是",
	"yes",
	"Yes",
	"YES",
	"y",
	"Y",
}

// Parser parser.
type Parser struct {
	FileName string
	// DataIndexOffset 数据索引偏移量,第一行可能为表头，则偏移量为1，如果有注释占用一行，则为2
	DataIndexOffset int
	// BoolTrueValues bool类型的true可选值 boolTrueValue
	BoolTrueValues []string
	// IsCheckEmpty 序列化到结构提的时候是否检测空值如果为空值则报错
	IsCheckEmpty bool
	// IsEmptyFunc 空值校验函数 可自定义空值
	IsEmptyFunc func(v string) bool
	// IsCoordinatesABS
	//    if true Cell.Coordinates is "A1"
	//    if false Cell.Coordinates is "$A$1"
	IsCoordinatesABS bool
	ExcelData        *Data
	// AllowFieldRepeat 允许表头字段重复
	AllowFieldRepeat bool
	currentSheetName string

	errsMap    map[string]error
	hasComment bool
	// fieldHeadRowIndex 表头行索引，第一行为表头，则索引为1
	fieldHeadRowIndex int
	excelFile         *excelize.File
}

// NewParser 传入文件名
// sheetIndex从1开始
//
//	ExcelTag excel tag:
//		ExcelField: map the excel head filed
//		ExcelDefault: if excel field is empty, use this default value
func NewParser(fileName string) *Parser {
	return &Parser{
		FileName:          fileName,
		DataIndexOffset:   1,
		fieldHeadRowIndex: 1,
		BoolTrueValues:    boolTrueValue,
		IsEmptyFunc: func(v string) bool {
			return v == ""
		},
		errsMap: make(map[string]error),
	}
}


reader.go
package excelstructure

import (
	"fmt"
	"reflect"
	"strconv"

	sliceutil "github.com/booyangcc/utils/sliceutil"
	"github.com/hashicorp/go-multierror"

	"github.com/xuri/excelize/v2"
)

// Parse parse.
func (p *Parser) Parse() (*Data, error) {
	fileName := p.FileName
	excelFile, err := excelize.OpenFile(fileName)
	if err != nil {
		return nil, NewError(fileName, "", "", err)
	}
	defer func() {
		if err1 := excelFile.Close(); err1 != nil {
			fmt.Println(err1.Error())
		}
	}()

	p.excelFile = excelFile
	if p.fieldHeadRowIndex < 1 {
		p.fieldHeadRowIndex = 1
	}
	if p.DataIndexOffset < 1 {
		p.DataIndexOffset = 1
	}

	excelData, err := p.getExcelFileData()
	if err != nil {
		return nil, NewError(fileName, "", "", err)
	}
	p.ExcelData = excelData
	return excelData, nil
}

// ReadWithSheetName parse with sheet index. start with 1
func (p *Parser) ReadWithSheetName(name string, output interface{}) error {
	excelData, err := p.Parse()
	if err != nil {
		return err
	}

	return p.readToStruct(name, excelData, output)
}

// Parser parse with sheet index 1
// output must be a pointer slice
// if the pointer field is pointer, and the value is empty ,the pointer field will be nil
func (p *Parser) Read(output interface{}) error {
	excelData, err := p.Parse()
	if err != nil {
		return err
	}
	return p.readToStruct("", excelData, output)
}

func (p *Parser) readToStruct(sheetName string, excelData *Data, output interface{}) (errs error) {
	if sliceutil.InSlice(sheetName, excelData.SheetList) {
		errs = multierror.Append(errs,
			NewError(p.FileName, "", fmt.Sprintf("sheetName %s", sheetName), ErrorSheetName))
		return
	}

	if sheetName == "" {
		if len(excelData.SheetList) == 0 {
			return
		}
		sheetName = excelData.SheetList[0]
	}

	p.currentSheetName = p.excelFile.GetSheetList()[0]

	sheetData := excelData.SheetNameData[sheetName]

	rv := reflect.ValueOf(output)
	if rv.Kind() == reflect.Ptr && rv.Elem().Kind() != reflect.Slice {
		errs = multierror.Append(errs,
			NewError(p.FileName, p.currentSheetName, "", ErrorInOutputType))
		return
	}

	sliceType := rv.Elem().Type()
	sliceElemType := sliceType.Elem()
	sliceElemStructType, err := getSliceElemType(p.FileName, p.currentSheetName, rv)
	if err != nil {
		errs = multierror.Append(errs, err)
		return
	}
	tagMap := parseFiledTagSetting(sliceElemStructType)

	arr := reflect.MakeSlice(sliceType, 0, 0)
	for i := range sheetData.Rows {
		out := reflect.New(sliceElemStructType)

		if err := p.parseRowToStruct(i, sheetData, out, tagMap); err != nil {
			p.appendError(errs, err)
			continue
		}
		if sliceElemType.Kind() == reflect.Ptr {
			arr = reflect.Append(arr, out)
		}
		if sliceElemType.Kind() == reflect.Struct {
			arr = reflect.Append(arr, out.Elem())
		}

	}
	rv.Elem().Set(arr)

	return
}

func (p *Parser) appendError(errs error, err error) {
	if errs == nil {
		return
	}
	if err == nil {
		return
	}
	if p.errsMap == nil {
		p.errsMap = make(map[string]error)
	}

	if _, ok := p.errsMap[err.Error()]; ok {
		return
	}
	errs = multierror.Append(errs, err)
	p.errsMap[err.Error()] = err
}

// parse row to struct by tag setting
func (p *Parser) parseRowToStruct(
	rowIndex int, sheetData *SheetData, ve reflect.Value, tagMap map[string]TagSetting,
) (err error) {
	if ve.Kind() != reflect.Ptr {
		return NewError(p.FileName, p.currentSheetName, fmt.Sprintf("row %d", rowIndex), ErrorTypePointer)
	}
	if !ve.IsValid() {
		return NewError(p.FileName, p.currentSheetName, fmt.Sprintf("row %d", rowIndex), ErrorFieldInvalid)
	}
	vek := reflect.Indirect(ve)
	createStruct := vek.Type()
	for i := 0; i < createStruct.NumField(); i++ {
		field := createStruct.Field(i)
		tag := field.Tag
		et := tag.Get(TagName)
		columnName, df, skip := field.Name, "", false
		if len(et) > 0 {
			val, ok := tagMap[field.Name]
			if ok {
				columnName = val.Column
				df = val.Default
				skip = val.Skip
			}
		}
		if skip {
			continue
		}
		cell, err1 := sheetData.GetCell(rowIndex, columnName)
		if err1 != nil {
			return err1
		}

		err = p.filedSetAll(ve.Elem().FieldByName(field.Name), cell, df)
		if err != nil {
			return err
		}

	}
	return err
}

func (p *Parser) filedSetAll(field reflect.Value, cell *Cell, defaultValue string) error {
	value := cell.Value
	if p.IsCheckEmpty && cell.IsEmpty {
		return NewError(p.FileName, p.currentSheetName, cell.Coordinates, ErrorFieldValueEmpty)
	}
	if value == "" {
		value = defaultValue
	}
	var err error
	switch field.Kind() {
	case reflect.Ptr:
		valType := field.Type()
		valElemType := valType.Elem()
		if field.CanSet() {
			realVal := field
			if realVal.IsNil() {
				realVal = reflect.New(valElemType)
				if value != "" {
					err = p.filedSet(realVal.Elem(), value, cell)
					field.Set(realVal)
				} else {
					field.Set(reflect.Zero(valType))
				}
			}
		} else {
			err = NewError(p.FileName, p.currentSheetName, cell.Coordinates, ErrorFieldNotSet)
		}
	default:
		if field.CanSet() {
			err = p.filedSet(field, value, cell)
		} else {
			err = NewError(p.FileName, p.currentSheetName, cell.Coordinates, ErrorFieldNotSet)
		}
	}
	if err != nil {
		return err
	}
	return nil
}

func (p *Parser) filedSet(field reflect.Value, value string, cell *Cell) error {
	if !field.IsValid() {
		fmt.Println(field.Kind())
		return NewError(p.FileName, p.currentSheetName, cell.Coordinates, ErrorFieldNotSet)
	}
	if !field.CanSet() {
		return NewError(p.FileName, p.currentSheetName, cell.Coordinates, ErrorFieldNotSet)
	}
	switch field.Kind() {
	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
		intValue, err := strconv.Atoi(value)
		if err != nil {
			return NewError(p.FileName, p.currentSheetName, cell.Coordinates, ErrorFieldNotMatch)
		}
		field.SetInt(int64(intValue))
	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
		uintValue, err := strconv.ParseUint(value, 10, 32)
		if err != nil {
			return NewError(p.FileName, p.currentSheetName, cell.Coordinates, ErrorFieldNotMatch)
		}
		field.SetUint(uintValue)
	case reflect.String:
		field.SetString(value)
	case reflect.Bool:
		field.SetBool(sliceutil.InSlice(value, p.BoolTrueValues))
	case reflect.Float32, reflect.Float64:
		floatValue, err := strconv.ParseFloat(value, 32)
		if err != nil {
			return NewError(p.FileName, p.currentSheetName, cell.Coordinates, ErrorFieldNotMatch)
		}
		field.SetFloat(floatValue)
	default:
		return NewError(p.FileName, p.currentSheetName, cell.Coordinates, ErrorFieldTypeNotSupport)
	}
	return nil
}
func (p *Parser) getExcelFileData() (*Data, error) {
	fileName := p.FileName
	sheetMap := p.excelFile.GetSheetMap()
	if nil == sheetMap || len(sheetMap) == 0 {
		return nil, NewError(fileName, "", "", ErrorNoSheet)
	}

	sheetIndexData := make(map[int]*SheetData)
	sheetNameData := make(map[string]*SheetData)
	for sheetIndex, sheetName := range sheetMap {
		rows, err := p.excelFile.GetRows(sheetName)
		if err != nil {
			return nil, NewError(fileName, sheetName, fmt.Sprintf("sheet index %d", sheetIndex), err)
		}
		if len(rows) == 0 {
			sheetData := &SheetData{
				SheetName:       sheetName,
				FileName:        fileName,
				DataIndexOffset: p.DataIndexOffset,
			}
			sheetIndexData[sheetIndex] = sheetData
			sheetNameData[sheetName] = sheetData
			continue
		}
		// 输入数据为excel直观的行数 从1开始
		sheetFields := rows[p.fieldHeadRowIndex-1]

		fieldValid := make([]string, 0)
		repeatFieldName := ""
		for _, fieldName := range sheetFields {
			if sliceutil.InSlice(fieldName, fieldValid) {
				repeatFieldName = fieldName
				break
			}
			fieldValid = append(fieldValid, fieldName)
		}
		// 检查是否有相同字段
		if !p.AllowFieldRepeat && repeatFieldName != "" {
			return nil, NewError(fileName, sheetName, fmt.Sprintf("sheet index %d", sheetIndex), ErrorFieldRepeat)
		}

		parseRows := make(map[int]map[string]*Cell, 0)
		for index, row := range rows {
			excelIndex := index + 1
			if excelIndex <= p.DataIndexOffset {
				continue
			}
			p.getRow(excelIndex, row, sheetFields, parseRows)
		}

		sheetData := &SheetData{
			RowTotal:        len(rows),
			DataTotal:       len(rows) - p.DataIndexOffset,
			SheetName:       sheetName,
			FileName:        fileName,
			Rows:            parseRows,
			FieldKeys:       sheetFields,
			DataIndexOffset: p.DataIndexOffset,
		}

		sheetIndexData[sheetIndex] = sheetData
		sheetNameData[sheetName] = sheetData
	}

	excelData := &Data{
		FileName:       fileName,
		SheetIndexData: sheetIndexData,
		SheetNameData:  sheetNameData,
		SheetTotal:     p.excelFile.SheetCount,
		SheetList:      p.excelFile.GetSheetList(),
	}

	return excelData, nil
}

func (p *Parser) getRow(rowIndex int, rawRow, sheetFields []string, rowsData map[int]map[string]*Cell) {
	rowData := make(map[string]*Cell, 0)
	// excel起始行为1，所以这里要+1
	for colIndex, fieldName := range sheetFields {
		cell := &Cell{
			RowIndex: rowIndex,
			ColIndex: colIndex + 1,
			Key:      fieldName,
		}
		var c string
		var err error
		if p.IsCoordinatesABS {
			c, err = excelize.CoordinatesToCellName(colIndex+1, rowIndex, true)
		} else {
			c, err = excelize.CoordinatesToCellName(colIndex+1, rowIndex)
		}
		if err != nil {
			cell.ErrMsg = c
		}
		// 当前列索引大于当前行长度，说明当前行数据不足，当前列为空
		if colIndex > len(rawRow)-1 {
			cell.IsEmpty = true
		} else {
			cell.Value = rawRow[colIndex]
			cell.IsEmpty = p.IsEmptyFunc(rawRow[colIndex])
		}
		cell.Coordinates = c
		rowData[fieldName] = cell
	}

	rowsData[rowIndex] = rowData
}


writer.go
package excelstructure

import (
	"fmt"
	"github.com/hashicorp/go-multierror"
	"github.com/xuri/excelize/v2"
	"reflect"
)

// Write struct to excel
func (p *Parser) Write(input interface{}) error {
	return p.writeToExcel(p.FileName, input)
}

func (p *Parser) writeToExcel(fileName string, input interface{}) (errs error) {
	rv := reflect.Indirect(reflect.ValueOf(input))
	if rv.Kind() != reflect.Slice {
		errs = multierror.Append(errs,
			NewError(p.FileName, p.currentSheetName, "", ErrorInOutputType))
		return
	}

	if rv.Len() == 0 {
		return
	}

	sliceElemStructType, err := getSliceElemType(p.FileName, p.currentSheetName, rv)
	if err != nil {
		errs = multierror.Append(errs, err)
		return
	}
	tagMap := parseFiledTagSetting(sliceElemStructType)

	excelFile := excelize.NewFile()

	sheetName := fmt.Sprintf("%ss", sliceElemStructType.Name())
	_, err = excelFile.NewSheet(sheetName)
	if err != nil {
		errs = multierror.Append(errs, err)
		return
	}
	p.currentSheetName = sheetName

	err = p.writeHead(excelFile, tagMap, sliceElemStructType)
	if err != nil {
		errs = multierror.Append(errs, err)
		return
	}
	err = p.writeData(excelFile, tagMap, rv)
	if err != nil {
		errs = multierror.Append(errs, err)
		return
	}

	err = excelFile.DeleteSheet(excelFile.GetSheetList()[0])
	if err != nil {
		errs = multierror.Append(errs, err)
		return
	}

	if err = excelFile.SaveAs(fileName); err != nil {
		errs = multierror.Append(errs,
			NewError(p.FileName, p.currentSheetName, "", err))
	}
	return
}

func (p *Parser) writeData(ef *excelize.File, tagMap map[string]TagSetting, rv reflect.Value) error {
	for i := 0; i < rv.Len(); i++ {
		elemValue := rv.Index(i)
		elemType := elemValue.Type()
		if elemType.Kind() == reflect.Ptr {
			elemType = elemType.Elem()
			elemValue = elemValue.Elem()
		}
		rowData := make([]interface{}, 0)
		fmt.Println(elemValue.String(), elemType.String(), elemType.NumField())
		for j := 0; j < elemType.NumField(); j++ {
			field := elemType.Field(i)
			filedTagSetting, ok := tagMap[field.Name]
			if !ok {
				filedTagSetting = TagSetting{
					Column: field.Name,
				}
			}
			if filedTagSetting.Column == "-" || filedTagSetting.Skip {
				continue
			}
			realElemValue := elemValue.Field(j).Interface()
			if elemValue.Field(j).Kind() == reflect.Ptr {
				realElemValue = elemValue.Field(j).Elem().Interface()
			}
			if elemValue.Field(j).IsZero() && filedTagSetting.Default != "" {
				realElemValue = filedTagSetting.Default
			}
			rowData = append(rowData, realElemValue)
		}

		coords, err := excelize.CoordinatesToCellName(1, p.DataIndexOffset+i+1)
		if err != nil {
			return err
		}
		err = ef.SetSheetRow(p.currentSheetName, coords, &rowData)
		if err != nil {
			return err
		}
	}
	return nil
}

func (p *Parser) writeHead(ef *excelize.File, tagMap map[string]TagSetting, sliceElemType reflect.Type) error {
	heads := make([]string, 0)
	comments := make([]string, 0)
	hasComment := false
	for _, tag := range tagMap {
		if tag.Comment != "" {
			hasComment = true
		}
	}

	for i := 0; i < sliceElemType.NumField(); i++ {
		field := sliceElemType.Field(i)
		filedTagSetting, ok := tagMap[field.Name]
		if !ok {
			filedTagSetting = TagSetting{
				Column: field.Name,
			}
		}
		if filedTagSetting.Column == "-" || filedTagSetting.Skip {
			continue
		}
		heads = append(heads, filedTagSetting.Column)
		if hasComment {
			comments = append(comments, filedTagSetting.Comment)
		}
	}

	err := ef.SetSheetRow(p.currentSheetName, "A1", &heads)
	if err != nil {
		return err
	}
	if hasComment {
		err = ef.SetSheetRow(p.currentSheetName, "A2", &comments)
		if err != nil {
			return err
		}
		p.DataIndexOffset = 2
	}
	p.hasComment = hasComment
	return nil
}

func getSliceElemType(fileName, currentSheetName string, rv reflect.Value) (reflect.Type, error) {
	sliceType := rv.Type().Elem()
	sliceElemType := sliceType.Elem()

	if sliceElemType.Kind() == reflect.Ptr {
		sliceElemType = sliceElemType.Elem()
	}
	if sliceElemType.Kind() != reflect.Struct {
		return nil, NewError(fileName, currentSheetName, "", ErrorSliceElemType)
	}

	return sliceElemType, nil
}


tag_setting.go
package excelstructure

import (
	"reflect"
	"strings"
)

const (
	// TagName tag name.
	TagName = "excel"
)

// TagSetting tag setting.
type TagSetting struct {
	// column
	Column string
	// 	Type string
	Type string
	// 	default value
	Default string
	Comment string
	Skip    bool
}

func parseTagSetting(str, sep, kvSep string) map[string]string {
	settings := map[string]string{}
	names := strings.Split(str, sep)

	for i := 0; i < len(names); i++ {
		values := strings.Split(names[i], kvSep)
		k := strings.TrimSpace(strings.ToLower(values[0]))

		if len(values) >= 2 {
			settings[k] = values[1]
		} else if k != "" {
			settings[k] = k
		}
	}

	return settings
}

func parseFiledTagSetting(sliceElemType reflect.Type) map[string]TagSetting {
	tagFiledMap := make(map[string]TagSetting)

	for i := 0; i < sliceElemType.NumField(); i++ {
		field := sliceElemType.Field(i)
		tag := field.Tag.Get(TagName)
		if _, ok := tagFiledMap[field.Name]; ok {
			continue
		}
		kvm := parseTagSetting(tag, ";", ":")
		tagFiled := TagSetting{
			Column:  kvm["column"],
			Type:    kvm["type"],
			Default: kvm["default"],
			Comment: kvm["comment"],
			Skip:    kvm["skip"] == "skip",
		}
		tagFiledMap[field.Name] = tagFiled
	}

	return tagFiledMap
}


examlple.go
package excelstructure

import (
	"fmt"
	"github.com/booyangcc/utils/convutil"
)

type Info struct {
	Name    string  `excel:"column:user_name;comment:person name"`
	Phone   *string `excel:"column:phone;comment:phone number"`
	Age     string  `excel:"column:age;"`
	Man     bool    `excel:"column:man;default:true"`
	Address string  `excel:"column:address;skip"` // skip this field
}

var (
	infos = []*Info{
		{
			Name:    "booyang",
			Phone:   convutil.String("123456789"),
			Age:     "18",
			Man:     true,
			Address: "beijing",
		},
		{
			Name:    "booyang1",
			Phone:   convutil.String("123456789"),
			Age:     "14",
			Man:     false,
			Address: "shanghai",
		},
	}
)

func TestWriteRead() {
	p := NewParser("./test_excel_file/test_write.xlsx")
	err := p.Write(infos)
	if err != nil {
		fmt.Println(err)
	}

	// because the struct field has comment tag, so the comment has been written to row 2, so when read, data offset is 2
	p.DataIndexOffset = 2
	var newInfo []*Info
	err = p.Read(&newInfo)
	if err != nil {
		fmt.Println(err)
	}
	fmt.Printf("oldInfo: %+v, newInfo: %+v", infos, newInfo)
}

func TestParse() {
	p := NewParser("./test_excel_file/test_write.xlsx")
	// because the struct field has comment tag, so the comment has been written to row 2, so when read, data offset is 2
	p.DataIndexOffset = 2
	excelData, err := p.Parse()
	if err != nil {
		fmt.Println(err)
	}

	s := excelData.SheetNameData["Infos"]
	row3UserName, err := s.GetStringValue(3, "user_name")
	if err != nil {
		fmt.Println(err)
	}

	row3age, err := s.GetIntValue(3, "age")
	if err != nil {
		fmt.Println(err)
	}
	fmt.Println(row3UserName, row3age)

	row4UserName, err := s.GetStringValue(4, "user_name")
	if err != nil {
		fmt.Println(err)
	}

	row4age, err := s.GetIntValue(4, "age")
	if err != nil {
		fmt.Println(err)
	}
	fmt.Println(row4UserName, row4age)
}
